### ISSUE TITLE
Correct config for functions and extend FunctionDefinitionBuilder

### ISSUE THREAD

**Issue by remixtedi** [betalgo/openai#317]
1. In general I'm getting the following error message when functions are provided in the request. Unrecognized request argument supplied: functions
Using the following configuration:
```
      "ApiVersion": "2023-06-01-preview",
      "DefaultModelId": "gpt-35-turbo-0613"

```

3.  I want toAdd a parameter with an "array" type and information about its items (types and descriptions) but I couldn't figure out how it is possible in the 7.1.2 beta version.
4.  It is not clear how to add a parameter with an "object" type.

Example:
![image](https://github.com/betalgo/openai/assets/7464545/c6418782-17e5-4d92-ae70-1814f561b1cf)

![image](https://github.com/openai/openai-node/assets/7464545/68d2a4b4-0198-42c1-8925-ad6933f80e6b)

Here is a sample code that I am using to test functions.

```c#
using OpenAI;
using OpenAI.Managers;
using OpenAI.ObjectModels.RequestModels;

var openAIClient = new OpenAIService(new OpenAiOptions()
{
    ApiKey = "YOUR API KEY",
    DeploymentId = "YOUR DEPLOYMENT ID",
    ResourceName = "YOUR RESOURCE NAME",
    ProviderType = ProviderType.Azure,
    ApiVersion = "2023-06-01-preview",
    DefaultModelId = "gpt-35-turbo-0613"
});
    
var func = new FunctionDefinitionBuilder("generate_brand_color", "Generates a brand color based on the brand name");
func.AddParameter("HexCode", "string", "Hex code of the brand color");

var chatCompletionCreateRequest = new ChatCompletionCreateRequest
{
    Messages = new ChatMessage[]
    {
        ChatMessage.FromSystem("You are helping a customer to generate brand color"),
        ChatMessage.FromUser("I need a brand color for my brand, brand name is Shambala and it is a clothing brand")
    },
    Model = "gpt-35-turbo-0613",
    Temperature = 0.9f,
    MaxTokens = 150,
    TopP = 1,
    FrequencyPenalty = 0,
    PresencePenalty = 0,
    FunctionCall = new Dictionary<string, string> { { "name", "generate_brand_color" } },
    Functions = new List<FunctionDefinition> { func.Build() }
};

var chatCompletion = await openAIClient.ChatCompletion.CreateCompletion(chatCompletionCreateRequest);
Console.WriteLine(chatCompletion.Choices.First().Message);
```

**Comment by rzubek** [comment]
Your second issue should be addressed in PR #320, which adds support for function arguments being arrays.

Not sure about the first and the last. I suspect the first is a matter of wrong API version? Functions only got released around 6/12.

**Comment by kayhantolga** [comment]
As @rzubek mentioned, some of our concerns will be addressed in the next release. I don't believe that Azure supports function calling yet. If I am wrong please provide related documentation.

**Comment by remixtedi** [comment]
@

> As @rzubek mentioned, some of our concerns will be addressed in the next release. I don't believe that Azure supports function calling yet. If I am wrong please provide related documentation.

Looks like they are working on it in **api-version=2023-07-01-preview** . Not sure how stable it is for now...

https://learn.microsoft.com/en-us/answers/questions/1323694/error-calling-function-when-using-model-gpt-35-tur

**Comment by kayhantolga** [comment]
I assume they will release a similar API with OpenAI. Defining object and array are now available in v7.1.3. More documentation will be coming soon. While we have a ticket for documentation and the rest of the issues are resolved, I will close this ticket. If any issues are still valid, please do not hesitate to reopen.
